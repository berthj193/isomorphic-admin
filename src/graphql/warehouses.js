import gql from 'graphql-tag';

// Currently there is an error with shipping_rates. After resolving the issue, include code below in warehouses query:
// shipping_rates {
//   id
//   name
//   country
//   currency
//   price_pence
// }
// options {
//   key
//   value
// }
export const getList = gql`
query {
  warehouses {
    id
  	name
    export_class
    created_at
    stores {
      id
    }
  }
}`;

export const getDetails = gql`
query getDetails($id: ID!){
  warehouses (id: $id) {
    id
  	name
    export_class
    created_at
    updated_at
    stores {
      id
      key
    }
  }
}`;

export const createWarehouse = gql`mutation createWarehouse(
  $name: String!, $export_class: String!
) {
  createWarehouse(
    name: $name,
    export_class: $export_class
  ) {
    id
  }
}`;

export const destroyWarehouse = gql`mutation destroyWarehouse(
  $id: ID!
) {
  destroyWarehouse(
    id: $id
  )
}`;

export const updateWarehouse = gql`mutation updateWarehouse(
  $id: ID!, $name: String!, $export_class: String!
) {
  updateWarehouse(
    id: $id,
    name: $name,
    export_class: $export_class
  ) {
    id
  }
}`;

export const getAutocompleteList = gql`query {
  warehouses {
    id
    name
  }
}`;
