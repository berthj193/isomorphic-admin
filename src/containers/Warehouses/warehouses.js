import React, { Component } from 'react';
import PropTypes from 'prop-types';
import columns from './columns';
import { getList, destroyWarehouse } from '../../graphql/warehouses';
import withStoresManager from '../../components/apolloProviders/withStoresManager';
import ItemList from '../../components/list/item-list';
import DeleteButton from '../../components/uielements/deleteButton';

export class Warehouses extends Component {
  static propTypes = {
    warehouses: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        name: PropTypes.string,
        export_class: PropTypes.string,
        created_at: PropTypes.string,
        stores: PropTypes.arrayOf(PropTypes.shape({
          id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        })),
        shipping_rates: PropTypes.arrayOf(PropTypes.shape({
          id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        })),
      }),
    ),
    error: PropTypes.string,
    mutate: PropTypes.func,
  };

  static defaultProps = {
    warehouses: [],
  };

  dataSource = () => this.props.warehouses.map(
    warehouse => ({
      key: warehouse.id,
      id: warehouse.id,
      name: warehouse.name,
      export_class: warehouse.export_class,
      created_at: warehouse.created_at,
      stores: warehouse.stores.length,
      shipping_rates: warehouse.shipping_rates,
    })
  );

  deleteWarehouse = id => e => {
    e.preventDefault();
    e.stopPropagation();
    this.props.mutate({
      mutation: destroyWarehouse,
      variables: {
        id,
      },
      messages: ['Warehouse removed successfully'],
    });
  };

  actionColumn = ({ id }) => (
    <ul className="table-action">
      <li>
        <DeleteButton
          tooltipText="Remove a warehouse"
          onConfirm={this.deleteWarehouse(id)}
        />
      </li>
    </ul>
  );

  columns = [
    ...columns,
    {
      title: 'Action',
      dataIndex: '',
      key: 'a',
      render: this.actionColumn,
    },
  ];

  render() {
    return (
      <ItemList
        data={this.dataSource()}
        error={this.props.error}
        columns={this.columns}
        title="Warehouses"
        buttons={[{
          title: 'New warehouse',
          link: '/dashboard/warehouses/new',
        }]}
        loading={!this.props.warehouses && !this.props.error}
      />
    );
  }
}

export default withStoresManager(Warehouses, { query: getList });
