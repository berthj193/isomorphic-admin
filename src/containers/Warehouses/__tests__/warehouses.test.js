import React from 'react';
import { Warehouses } from '../warehouses';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('../../../helpers/client/storeManagerClient');

describe('<Warehouses />', () => {
  const getComponent = (props = {}) => shallow(<Warehouses {...props} />);
  const mutate = jest.fn();
  const mockWarehouses = [{
    id: 1,
    name: 'Warehouse',
    export_class: 'Foo',
    stores: [],
  }];

  beforeEach(() => {
    jest.resetAllMocks();
    mutate.mockReturnValue(Promise.resolve({ foo: 'bar' }));
  });

  it('should render properly', () => {
    expect(toJson(getComponent())).toMatchSnapshot();
  });

  it('should map and pass orders to table component', () => {
    const mockOrders = [
      {
        id: '1',
        name: 'Test',
        export_class: 'exportClass',
        created_at: '2018-05-31 15:43:15 UTC',
      },
    ];
    const component = getComponent();
    component.instance().setState({
      orders: mockOrders,
    });

    expect(toJson(component.update())).toMatchSnapshot();
  });

  it('dataSource method should return mapped object', () => {
    const component = getComponent({ warehouses: mockWarehouses });
    expect(component.instance().dataSource().length).toBe(1);
  });

  it('should execute mutate function on delete', () => {
    const wrapper = getComponent({ mutate });
    const event = {
      preventDefault: () => null,
      stopPropagation: () => null,
    };
    wrapper.instance().deleteWarehouse(1)(event);
    expect(mutate).toBeCalled();
  });

  it('action column should match snapshot', () => {
    expect(toJson(getComponent().instance().actionColumn({ id: 1 }))).toMatchSnapshot();
  });
});
