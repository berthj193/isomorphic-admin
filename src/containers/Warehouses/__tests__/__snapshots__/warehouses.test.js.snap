// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Warehouses /> action column should match snapshot 1`] = `undefined`;

exports[`<Warehouses /> should map and pass orders to table component 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/warehouses/new",
        "title": "New warehouse",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "Name",
      },
      Object {
        "dataIndex": "export_class",
        "key": "export_class",
        "title": "Export class",
      },
      Object {
        "dataIndex": "created_at",
        "key": "created_at",
        "title": "Created at",
      },
      Object {
        "dataIndex": "stores",
        "key": "stores",
        "title": "Stores",
      },
      Object {
        "dataIndex": "shipping_rates",
        "key": "shipping_rates",
        "title": "Shipping rates",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  loading={false}
  title="Warehouses"
/>
`;

exports[`<Warehouses /> should render properly 1`] = `
<withRouter(ItemList)
  buttons={
    Array [
      Object {
        "link": "/dashboard/warehouses/new",
        "title": "New warehouse",
      },
    ]
  }
  columns={
    Array [
      Object {
        "dataIndex": "id",
        "key": "id",
        "title": "ID",
      },
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "Name",
      },
      Object {
        "dataIndex": "export_class",
        "key": "export_class",
        "title": "Export class",
      },
      Object {
        "dataIndex": "created_at",
        "key": "created_at",
        "title": "Created at",
      },
      Object {
        "dataIndex": "stores",
        "key": "stores",
        "title": "Stores",
      },
      Object {
        "dataIndex": "shipping_rates",
        "key": "shipping_rates",
        "title": "Shipping rates",
      },
      Object {
        "dataIndex": "",
        "key": "a",
        "render": [Function],
        "title": "Action",
      },
    ]
  }
  data={Array []}
  loading={false}
  title="Warehouses"
/>
`;
